### PR TITLE
feat: new catalog endpoint that generates a diff of a catalog's content

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
## Description

Diff is based on provided keys. 

context- we need a new endpoint to generate 3 buckets of data based on a diff of a catalog's content and a provided list of keys. The buckets signify which provided keys are listed under the catalog, which were not, and what contents are listed but not provided.

## Ticket Link

Link to the associated ticket
[Jira ticket](https://openedx.atlassian.net/browse/ENT-5044)

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

File: enterprise_catalog/apps/api/v1/views.py [hunk 1]
```diff
@@ -164,6 +164,74 @@ def contains_content_items(self, request, uuid, course_run_ids, program_uuids, *
...
+            enterprise_catalog = self.get_object()
+            return str(enterprise_catalog.enterprise_uuid)
+        return None
+
+    @action(detail=True)
+    def post(self, request, **kwargs):
+        course_keys = request.data.get('course_keys')
+        return self.catalog_diff(course_keys)
+
+    @method_decorator(require_at_least_one_query_parameter('course_keys'))
```
Comment (binodpant) [inline enterprise_catalog/apps/api/v1/views.py]: not clear why we need a GET at all, but since you set this up already we can go with it!

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 2]
```diff
@@ -889,6 +889,90 @@ def _get_contains_content_base_url(self, enterprise_uuid=None):
...
+        """
+        Test that the generate_diff endpoint will return under the `items_found` bucket all content keys within the
+        catalog that were provided.
+        """
+        content = ContentMetadataFactory()
+        self.add_metadata_to_catalog(self.enterprise_catalog, [content])
+
+        url = self._get_generate_diff_base_url()
+        response = self.client.post(url, data={'course_keys': [content.content_key]})
+        assert response.data.get('items_found')[0].get('course_key') == content.content_key
```
Comment (binodpant) [inline enterprise_catalog/apps/api/v1/tests/test_views.py]: think it is better to use multiple values in an array here just to be stricter for the testing

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 3]
```diff
@@ -889,6 +889,90 @@ def _get_contains_content_base_url(self, enterprise_uuid=None):
...
+    def test_generate_diff_returns_content_items_not_found(self):
+        """
+        Test that the generate_diff endpoint will return all content keys provided that were not found under the catalog
+        under the `items_not_found` bucket.
+        """
+        key = 'bad+key'
+        key2 = 'bad+key2'
+        url = self._get_generate_diff_base_url()
+        response = self.client.post(url, data={'course_keys': [key, key2]})
+        assert response.data.get('items_not_found')[0] == {key, key2}
```
Comment (binodpant) [inline enterprise_catalog/apps/api/v1/tests/test_views.py]: can we also test the other return values here items_found and items_not_included (will they be empty?) actually would do that for all tests for more complete testing

File: enterprise_catalog/apps/api/v1/tests/test_views.py [hunk 4]
```diff
@@ -889,6 +889,90 @@ def _get_contains_content_base_url(self, enterprise_uuid=None):
...
+        Test that GET requests to the generate_diff endpoint behave the same as POST requests.
+        """
+        content = ContentMetadataFactory()
+        content2 = ContentMetadataFactory()
+        self.add_metadata_to_catalog(self.enterprise_catalog, [content, content2])
+        url = self._get_generate_diff_base_url()
+        response = self.client.post(url, data={'course_keys': [content.content_key, 'bad+key']})
+        assert response.status_code == 200
+        response_data = response.data
+        assert response_data.get('items_not_found')[0] == {'bad+key'}
```
Comment (binodpant) [inline enterprise_catalog/apps/api/v1/tests/test_views.py]: would recommend testing with more than one value here for each category

File: enterprise_catalog/apps/api/v1/views.py (hunk 1 above)
Comment (alex-sheehan-edx) [reply enterprise_catalog/apps/api/v1/views.py]: to allow for flexibility with consumers, not totally necessary but one or two people expressed interest so I added

**Review by binodpant — APPROVED** [review]
gtg!

File: enterprise_catalog/apps/api/v1/views.py [hunk 5]
```diff
@@ -164,6 +164,77 @@ def contains_content_items(self, request, uuid, course_run_ids, program_uuids, *
...
+            if len(content_keys) > self.MAX_GET_CONTENT_KEYS:
+                return Response(
+                    f'catalog_diff GET requests supports up to {self.MAX_GET_CONTENT_KEYS}. If more content keys '
+                    f'required, please use a POST body.',
+                    status=HTTP_400_BAD_REQUEST
+                )
+
+        return self.catalog_diff(content_keys)
+
+    def catalog_diff(self, course_keys):
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/api/v1/views.py]: nit: `content_keys` vs. `course_keys` here and elsewhere in docstrings, etc.

File: enterprise_catalog/apps/api/v1/views.py [hunk 6]
```diff
@@ -164,6 +164,77 @@ def contains_content_items(self, request, uuid, course_run_ids, program_uuids, *
...
+        if self.kwargs.get('uuid'):
+            enterprise_catalog = self.get_object()
+            return str(enterprise_catalog.enterprise_uuid)
+        return None
+
+    @action(detail=True)
+    def post(self, request, **kwargs):
+        course_keys = []
+        if request.data:
+            course_keys = request.data.getlist('content_keys')
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/api/v1/views.py]: nit: "content_keys"

File: enterprise_catalog/apps/catalog/models.py [hunk 7]
```diff
@@ -203,6 +203,50 @@ def enterprise_customer(self):
         """
         return EnterpriseCustomerDetails(self.enterprise_uuid)
 
+    def get_catalog_content_diff(self, content_keys):
+        """
+        Generate a catalog diff based on a provided list of content keys and what currently exists with the catalog's
+        metadata.
+
+        Returns:
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/catalog/models.py:214]: nit: Maybe also include "Arguments" to describe the `content_keys` arg?

File: enterprise_catalog/apps/catalog/models.py [hunk 8]
```diff
@@ -203,6 +203,50 @@ def enterprise_customer(self):
...
+        items_not_included = []
+        items_found = []
+        items_not_found = set()
+
+        # cannot determine if specified content keys are part of catalog when catalog
+        # query doesn't exist.
+        if not self.catalog_query:
+            return [items_not_found], items_not_included, items_found
+
+        content_keys = set(content_keys)
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/catalog/models.py]: nit: might recommend remaining `content_keys` variable to `distinct_content_keys` or similar so that we're not reusing the same name as the method argument.

File: enterprise_catalog/apps/catalog/models.py [hunk 9]
```diff
@@ -203,6 +203,50 @@ def enterprise_customer(self):
...
+        for content in self.content_metadata.all().values('modified', 'content_key'):
+            found_content_keys.add(content.get('content_key'))
```
Comment (adamstankiewicz) [inline enterprise_catalog/apps/catalog/models.py]: nit: `content.get('content_key')` is re-used in several places here. Maybe create a variable to store this value without needing to call `.get` several times.

**Review by adamstankiewicz — APPROVED** [review]
Just a few minor nits, otherwise looks good to me!
